### PR TITLE
Mc address schema investigation

### PR DESCRIPTION
**Pull request by micahchiang**
## Description
There are currently 3 forms in `vets-website` that rely on an address schema utility [here](https://github.com/department-of-veterans-affairs/vets-website/blob/master/src/applications/disability-benefits/686c-674/config/address-schema.js). 

The problem with this is two of the forms, VRE 28-1900 and 28-8832, then rely on the schema used for the 686c-674 in `vets-json-schema`. The goal of this pull request is to create a new address schema utility at the root of `applications/vre` which is form agnostic. The intent is to eventually migrate the 686c-674 onto this new utility as well, so all three forms rely on an address schema independent from any one form. 

This address pattern is different from another commonly used pattern in `vets-website` found [here](https://github.com/department-of-veterans-affairs/vets-website/blob/master/src/platform/forms/definitions/address.js) because it includes a checkbox to indicate whether or not a user lives on a US military base outside of the United States. 

The state of this checkbox is used to manage the individual schemas for `country`, `state`, and `city`. This same pattern can be seen in a veteran's profile under the address block at `/profile/personal-information`, and found in vets-website [here](https://github.com/department-of-veterans-affairs/vets-website/blob/master/src/platform/user/profile/vet360/components/AddressField/address-schemas.js). 

It's important to note that a replication of this pattern was necessary, as the version used in the profile includes a hook, `updateUiSchema`, that isn't [called as part of the data-update flow in a config powered application](https://github.com/department-of-veterans-affairs/vets-website/blob/master/src/platform/forms-system/src/js/state/helpers.js#L357).

Corresponding schema object would look like:
<details><summary>Address Schema Object</summary>

```
export const addressSchema = {
  type: 'object',
  properties: {
    isMilitary: {
      type: 'boolean',
    },
    'view:livesOnMilitaryBaseInfo': {
      type: 'object',
      properties: {},
    },
    country: {
      type: 'string',
      enum: countries.map(country => country.value),
      enumNames: countries.map(country => country.label),
    },
    addressLine1: {
      type: 'string',
      minLength: 1,
      maxLength: 100,
      pattern: '^.*\\S.*',
    },
    addressLine2: {
      type: 'string',
      minLength: 1,
      maxLength: 100,
      pattern: '^.*\\S.*',
    },
    addressLine3: {
      type: 'string',
      minLength: 1,
      maxLength: 100,
      pattern: '^.*\\S.*',
    },
    city: {
      type: 'string',
    },
    state: {
      type: 'string',
    },
    postalCode: {
      type: 'string',
    },
  },
};
```

</details>

## Testing done
- local on the address page of the non-production 28-8832 application

## Screenshots
<details><summary>Empty form state</summary>

Screen Shot 2020-10-08 at 12 49 35 PM
</details>

<details><summary>Domestic US address</summary>

Screen Shot 2020-10-08 at 12 50 28 PM
</details>

<details><summary>US military address</summary>

Screen Shot 2020-10-08 at 12 52 37 PM
</details>

<details><summary>Foreign address</summary>

Screen Shot 2020-10-08 at 12 51 03 PM
</details>

## Acceptance criteria
- [x] New address schema UI works as expected on an address page inside a form system application. 

## Definition of done
- [ ] Events are logged appropriately
- [ ] Documentation has been updated, if applicable
- [x] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)
- [x] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs

### CODE REVIEW

File: src/applications/vre/definitions/profileAddress.js
Comment (micahchiang) [inline src/applications/vre/definitions/profileAddress.js:24]: In a follow up pull request, this will be moved into vets-json-schema and live there. Each form using this pattern will have this address schema declared in their individual form json-schema file.

**Comment by micahchiang** [comment]
@Mottie @erikphansen @cvalarida @cohnjesse - Requesting code review on this as a draft because I would like any feedback or constructive criticisms on how this could be improved before merging. Thanks!

File: src/applications/vre/definitions/profileAddress.js
Comment (cvalarida) [inline src/applications/vre/definitions/profileAddress.js:43]: Might be worth pulling this into a constant for reuse. :man_shrugging:

File: src/applications/vre/definitions/profileAddress.js
Comment (cvalarida) [inline src/applications/vre/definitions/profileAddress.js:113]: I'd recommend renaming `callback` to `uiRequiredCallback` or something a little more descriptive.

File: src/applications/vre/definitions/profileAddress.js
Comment (cvalarida) [inline src/applications/vre/definitions/profileAddress.js:114]:
```suggestion
  // Used when addresses are nested in an array and need to be accessible.
```

File: src/applications/vre/definitions/profileAddress.js
Comment (cvalarida) [inline src/applications/vre/definitions/profileAddress.js:134]: I'd recommend putting a JSDoc comment on this function to make it easier for future maintainers to understand _why_ this is necessary.

File: src/applications/vre/definitions/profileAddress.js
Comment (cvalarida) [inline src/applications/vre/definitions/profileAddress.js:115]: - `addressUiSchema` is used in one place—the claimant address page
- `path` is only ever `'claimantAddress'`
- `getPath` is only ever called with `path`
- `insertArrayIndex` is only ever called from `getPath`

Given the above, is `getPath` and `insertArrayIndex` ever needed? This feels like it might be over-engineering for now. Do you know that it will be needed in the future?

Particularly, the `'[INDEX]'` magic would make me a little nervous if I had to maintain this later.

File: src/applications/vre/definitions/profileAddress.js
Comment (cvalarida) [inline src/applications/vre/definitions/profileAddress.js:180]: Same as on the `updateSchema` callback, I'd recommend writing a JSDoc for this to help future maintainers understand the _why_ behind this.

File: src/applications/vre/definitions/profileAddress.js
Comment (cvalarida) [inline src/applications/vre/definitions/profileAddress.js:164]: :thinking: Does this show for the `maxLength` validation error? I thought this was only for the `pattern` violation (which in this case would be the `/^.*\\S.*/` regex).

File: src/applications/vre/definitions/profileAddress.js
Comment (cvalarida) [inline src/applications/vre/definitions/profileAddress.js:247]: Same story with the JSDoc.

File: src/applications/vre/definitions/profileAddress.js
Comment (cvalarida) [inline src/applications/vre/definitions/profileAddress.js:253]: I've always found dropping a comment on regex patterns helpful for understanding the intent and reason later.

File: src/applications/vre/definitions/profileAddress.js
Comment (Mottie) [inline src/applications/vre/definitions/profileAddress.js:14]: Most of the `ADDRESS_DATA` values are pretty much the same as the values in `vets-json-schema/dist/constants.json`; I'm not sure if this can be optimized.

File: src/applications/vre/definitions/profileAddress.js
Comment (Mottie) [inline src/applications/vre/definitions/profileAddress.js:68]: This is already defined: https://github.com/department-of-veterans-affairs/vets-json-schema/blob/master/src/common/address.js#L252-L256

File: src/applications/vre/definitions/profileAddress.js
Comment (Mottie) [reply src/applications/vre/definitions/profileAddress.js:164]: Also, I the messaging guidelines say ([ref](https://github.com/department-of-veterans-affairs/vets.gov-team/blob/master/Products/Platform/Design%20System/Guidelines/Error%20handling/Dictionary.md#task-completion-successfailure)):

> Please fill in a valid [address/email address/phone number].

File: src/applications/vre/definitions/profileAddress.js
Comment (Mottie) [reply src/applications/vre/definitions/profileAddress.js:253]: I totally understand the reason for that regex pattern 

ghe-emoji

File: src/applications/vre/definitions/profileAddress.js
Comment (Mottie) [inline src/applications/vre/definitions/profileAddress.js:123]: Maybe I'm confused... doesn't an IIFE that returns an object, do the same thing as just returning the object? Is it necessary?

File: src/applications/vre/28-8832/tests/config/claimant-information/claimant-address.unit.spec.jsx
Comment (micahchiang) [reply src/applications/vre/28-8832/tests/config/claimant-information/claimant-address.unit.spec.jsx:16]: You're right. In a follow up PR, the updated `addressSchema` will be added to `vets-json-schema` and included in the definitions for this form. The mismatch will be resolved once that change is included.

File: src/applications/vre/definitions/profileAddress.js
Comment (micahchiang) [reply src/applications/vre/definitions/profileAddress.js:113]: Good idea, will update.

File: src/applications/vre/definitions/profileAddress.js
Comment (micahchiang) [reply src/applications/vre/definitions/profileAddress.js:114]: 🤦

File: src/applications/vre/definitions/profileAddress.js
Comment (micahchiang) [reply src/applications/vre/definitions/profileAddress.js:115]: It's true, `addressUiSchema` is only used in the claimant address page right now. However, in the future a goal is to migrate our 686c-674 form into using this utility. In that form, there are instances where an address may be nested in an array. In that case:
- `path` may be equal to something like `childrenToAdd[INDEX].childAddressInfo.address`
- `getPath()` is then needed because it handles the `index` argument that is included in `updateSchema`
- `insertArrayIndex` is used because I couldn't think of a better way to handle multiple different index values for an array. I.e. hard coding something like `childrenToAdd[1].childAdressInfo.address` is problematic for obvious reasons. 

I could include some of the reasoning above in a jsdoc comment 🤔

File: src/applications/vre/definitions/profileAddress.js
Comment (micahchiang) [reply src/applications/vre/definitions/profileAddress.js:164]: 👀  had no idea that messaging guideline existed!! I've been winging a lot of the error messages =\ not a good look on my part. 

Good thought, Chris. I'll go back and take a look at this.

File: src/applications/vre/definitions/profileAddress.js
Comment (micahchiang) [reply src/applications/vre/definitions/profileAddress.js:68]: excellent

File: src/applications/vre/definitions/profileAddress.js
Comment (micahchiang) [reply src/applications/vre/definitions/profileAddress.js:123]: You're right, it isn't necessary 👍 I think I thought it might have been, in order to keep some things in scope, but now that you mention it, it works just fine without the IIFE. 

_time to go back and re-read documentation on IIFEs_

File: src/applications/vre/definitions/profileAddress.js
Comment (micahchiang) [reply src/applications/vre/definitions/profileAddress.js:164]: Update: @cvalarida is right. This pattern error message maps to the regex used, not `maxLength`.

File: src/applications/vre/definitions/profileAddress.js
Comment (Mottie) [inline src/applications/vre/definitions/profileAddress.js:18]: This comment is outdated now

File: src/applications/vre/definitions/profileAddress.js
Comment (cohnjesse) [inline src/applications/vre/definitions/profileAddress.js:27]: I have seen a lot of different implementations for reg ex patterns used throughout the VA.gov stack. I wonder if we can audit all of these and come up with one set that can be used everywhere?

File: src/applications/vre/definitions/profileAddress.js
Comment (cohnjesse) [inline src/applications/vre/definitions/profileAddress.js:46]: Prefill for addresses uses `street` and `street2` to prefill the address streets, I wonder if it would be better to use that since it would make prefill work out of the box without re-mapping anything? https://github.com/department-of-veterans-affairs/vets-api/blob/7360d520f2dfdd7b5a902139eaea44cef967b103/app/models/form_profile.rb#L51

**Review by cohnjesse — APPROVED** [review]
LGTM, left a couple of comments but no show stoppers

File: src/applications/vre/definitions/profileAddress.js
Comment (micahchiang) [reply src/applications/vre/definitions/profileAddress.js:46]: Oh good catch, I'll update. It brings up another question though - we offer 3 address lines, but only support prefill for two. Should a third street address line be added to prefill?

Comment (Mottie) [reply src/applications/vre/definitions/profileAddress.js]: I think we should support 3 address lines since vets profile has 3 lines. Not all endpoints accept 3 lines, so maybe in that case combine line 2 & 3 and truncate it, but that would really depend on the endpoint requirements.

**Pull request merged**
